staging/src/k8s.io/apiserver/pkg/server/signal.go

```diff
@@ -0,0 +1,59 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package server
+
+import (
+	"os"
+	"os/signal"
+)
+
+// SetupSignalHandler registered for SIGTERM and SIGINT. A stop channel is returned
+// which is closed on one of these signals. An exitCode channel is returned for the program
+// to send the desired code when the shutdown is finished. If exitCodeCh is written to
+// before stopCh is closed, the stopCh is closed by the handler.
+//
+// Usage:
+//   stopCh, exitCodeCh := SetupSignalHandler()
+//   exitCode := 1
+//   defer func() { exitCodeCh <- exitCode }()
+//   if err := server.Run(stopCh); err == nil {
+//		exitCode = 0
+//   }
+func SetupSignalHandler() (stopCh <-chan struct{}, exitCodeCh chan<- int) {
+	stop := make(chan struct{})
+	exit := make(chan int, 1) // do not block
+	c := make(chan os.Signal, 2)
+	signal.Notify(c, shutdownSignals...)
+	go func() {
+		select {
+		case <-c:
+			close(stop)
+			select {
+			case code := <-exit:
+				os.Exit(code)
+			case <-c:
+				// second signal. Exit directly.
+				os.Exit(1)
+			}
+		case code := <-exit:
+			close(stop)
+			os.Exit(code)
```

Comment:
Doesn't that create a race between main function returning 0 and the `os.Exit` call?